### DISCUSSION TITLE
Showing Typescript stacktrace Node.js

### DISCUSSION THREAD

**Discussion by iliasmentz** [getsentry/support#136] [Support]
Hi 👋 ,

I have successfully uploaded my source maps into Sentry for my Typescript Node app (Express).
However, I am still seeing the errors on Sentry as readable JS, not TS.

Can someone help me understand what I did wrong?
tsconfig.json
```json
{
   "compilerOptions": {
      "lib": ["es5","es6"],
      "target": "es6",
      "module": "commonjs",
      "moduleResolution": "node",
      "sourceMap": true,
      "sourceRoot": "/",
      "inlineSources": true,
      "noEmitHelpers": true,
      "allowJs": true,
      "rootDir": "./src",
      "outDir": "./build",                          
      "strict": true,         
      "noImplicitAny": false,
      "esModuleInterop": true,
      "resolveJsonModule": true,
      "experimentalDecorators": true,
      "emitDecoratorMetadata": true,
      "strictPropertyInitialization": false,
      "removeComments": false,
      "declaration":true
   },
}
```

Alert.ts (where I initialize sentry)
```ts
export default class Alert {
  static init(app: express.Router) {
    Sentry.init({
      release: process.env.SENTRY_RELEASE,
      dist: process.env.SENTRY_RELEASE,
      dsn: process.env.SENTRY_DSN,
      integrations: [
        new RewriteFrames({
          root: path.resolve(__dirname),
        }),
        new Sentry.Integrations.Http({ tracing: true }),
        new Tracing.Integrations.Express({ app: app }),
      ],
      environment: process.env.NODE_ENV,
      tracesSampleRate: 1.0,
    });
  }

  static sendErrorToSentry(error: Error) {
    Sentry.captureException(error);
  }
}
```

sentry_release.sh:
```sh

npm install typescript@4.5.5
npm install
npm run build --production

install_sentry_cli () {
  # Check if the URL is reachable and install the CLI
  echo "Fetching Sentry CLI"
  [[ -z $(curl -sL https://sentry.io/get-cli/) ]] && echo "Could not fetch Sentry CLI" || curl -sL https://sentry.io/get-cli/ | bash
}

# Check if the CLI exists and install if necessary
[ -e "/usr/local/bin/sentry-cli" ] && echo 'Sentry CLI exists. Skipping installation' || install_sentry_cli

echo "Sentry: logging in"
sentry-cli login --auth-token "$SENTRY_AUTH_TOKEN"

# Read version from environment or get the last commit SHA1 via Sentry CLI
[[ ! -z "$RELEASE_VERSION" ]] && version=$RELEASE_VERSION || version=$(sentry-cli releases propose-version)

echo "Sentry: creating release \"$version\""
sentry-cli releases new "$version"

echo "Sentry: associating commits"
sentry-cli releases set-commits "$version" --commit "my-org/my-branch@HEAD"

echo "Sentry: uploading source maps"
sentry-cli releases files "$version" upload-sourcemaps --ext js --ext map $PWD/build/** --dist "$version"

echo "Sentry: finalizing release \"$version\""
sentry-cli releases finalize "$version"

echo "Sentry: creating deploy \"$version\""
name=$(TZ=GMT date +"%Y%m%d%H%M%S")
sentry-cli releases deploys "$version" new
```

the issue on sentry looks like this:
<img width="607" alt="image" src="https://user-images.githubusercontent.com/20126610/199793967-73abd0ea-f307-41d1-934e-b9ee3a2c5e23.png">

the source map explain output is this:
<img width="669" alt="image" src="https://user-images.githubusercontent.com/20126610/199793881-eed38871-8cc3-4af6-976d-a272d392a0ab.png">

I could use some help 😅  thanks in advance!

**Comment by AbhiPrasad** [comment]
@iliasmentz could you try using `process.cwd()` instead of `path.resolve(__dirname)`?

```ts
const cwd = process.cwd();

export default class Alert {
  static init(app: express.Router) {
    Sentry.init({
      release: process.env.SENTRY_RELEASE,
      dist: process.env.SENTRY_RELEASE,
      dsn: process.env.SENTRY_DSN,
      integrations: [
        new RewriteFrames({
          root: cwd,
        }),
        new Sentry.Integrations.Http({ tracing: true }),
        new Tracing.Integrations.Express({ app: app }),
      ],
      environment: process.env.NODE_ENV,
      tracesSampleRate: 1.0,
    });
  }

  static sendErrorToSentry(error: Error) {
    Sentry.captureException(error);
  }
}
```

**Reply by iliasmentz** [reply]
Thanks for the reply @AbhiPrasad !
Unfortunately that only broke the explain as now it says:
```
✔ Fetched data for event: 941c163a13ba441bbb43b25baaa8faae
✔ Event has release name: 2b8e80a489b75d95ae3db49163a225e4747f27cf
✔ Event has a valid exception present
✔ Event has a valid stacktrace present
✖ Event exception stacktrace selected frame (0) has incorrect abs_path (valid url is required). Found /usr/app/build/services/venue/middleware/isAuthenticated.js
```
let me know if you have any more ideas 😅

**Reply by souredoutlook** [reply]
Hey @iliasmentz - it might be helpful at this time to know what project this is for (if you're using SAAS). Can you share the project ID (not the slug) - you can find it in your DSN like so:

```
{PROTOCOL}://{PUBLIC_KEY}:{SECRET_KEY}@{HOST}{PATH}/{PROJECT_ID}
```

**Reply by iliasmentz** [reply]
Hi sure, thanks for the quick reply.
Regarding the project, this is a monorepo that contains an API and two WebSockets (3 sentry projects in total).
They serve a mobile app and it's written using Typescript, Node, Express.js/Socket.io.

The project ids are:
- API: `6719021`
- Socket 1: `6719038`
- Socket 2: `6719023`

**Reply by souredoutlook** [reply]
Great thanks! I think we're almost there. Would you be able to also share the event ID used in the first screenshot you shared with results from the source-maps-explain command?

**Reply by iliasmentz** [reply]
of course, the event id is `2fec6a6ada8a47f2a6d959f0c5753b06`

**Reply by souredoutlook** [reply]
Hey @iliasmentz - after further review there are some obvious things that can be changed (like omiting the dist value in but Alerts.ts and sentry_release.sh).

But other aspects of your issue are much less clear, for example, your raw stacktrace has files from both build and node_modules and RewriteFrames does not appear to be functioning as expected.

I'll needs logs from your app with the debug option enabled. Given the PII concerns around this I'd like to open up a private Zendesk ticket.

Can I use the email we have on file for your user in the Sentry org? Let me know how you want to proceed.

**Reply by iliasmentz** [reply]
yeah, that makes sense. let's transfer that to zendesk. yes please, use the email in the sentry org. thank you

**Comment by souredoutlook** [comment]
Hi @iliasmentz - I've picked this up in Zendesk #74910. I am also going to open an issue from here to track both deliverables in a GitHub project: CLI output, and your sourcemaps issue.
